subsys/usb/host/usbh_shell.c

```diff
@@ -0,0 +1,541 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/usb/usbh.h>
+
+#include "usbh_internal.h"
+#include "usbh_ch9.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbh_shell, CONFIG_USBH_LOG_LEVEL);
+
+#define FOOBAZ_VREQ_OUT		0x5b
+#define FOOBAZ_VREQ_IN		0x5c
+
+USBH_CONROLLER_DEFINE(uhs_ctx, DEVICE_DT_GET(DT_NODELABEL(zephyr_uhc0)));
+
+const static struct shell *ctx_shell;
+
+static void print_dev_desc(const struct shell *sh,
+			   const struct usb_device_descriptor *const desc)
+{
+	shell_print(sh, "bLength\t\t\t%u", desc->bLength);
+	shell_print(sh, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(sh, "bcdUSB\t\t\t%x", desc->bcdUSB);
+	shell_print(sh, "bDeviceClass\t\t%u", desc->bDeviceClass);
+	shell_print(sh, "bDeviceSubClass\t\t%u", desc->bDeviceSubClass);
+	shell_print(sh, "bDeviceProtocol\t\t%u", desc->bDeviceProtocol);
+	shell_print(sh, "bMaxPacketSize0\t\t%u", desc->bMaxPacketSize0);
+	shell_print(sh, "idVendor\t\t%x", desc->idVendor);
+	shell_print(sh, "idProduct\t\t%x", desc->idProduct);
+	shell_print(sh, "bcdDevice\t\t%x", desc->bcdDevice);
+	shell_print(sh, "iManufacturer\t\t%u", desc->iManufacturer);
+	shell_print(sh, "iProduct\t\t%u", desc->iProduct);
+	shell_print(sh, "iSerial\t\t\t%u", desc->iSerialNumber);
+	shell_print(sh, "bNumConfigurations\t%u", desc->bNumConfigurations);
+}
+
+static int bazfoo_request(struct usbh_contex *const ctx,
+			  struct uhc_transfer *const xfer,
+			  int err)
+{
+	const struct device *dev = ctx->dev;
+
+	shell_info(ctx_shell, "host: transfer finished %p, err %d", xfer, err);
+
+	while (!k_fifo_is_empty(&xfer->done)) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN &&
+			    buf->len == sizeof(struct usb_device_descriptor)) {
```

Comment:
This will treat any descriptor as device descriptor if the size matches. This can be really confusing.

Reply:
Yes, will try to improve it until tomorrow.